docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,442 @@
+     * @param[out] params_size The size of the evidence parameter.
+     * @retval OE_OK on success.
+     * @retval other appropriate error code.
+     */
+    oe_result_t (*get_evidence_parameters(
+        uint8_t** params,
+        size_t* params_size);
+};
+```
+
+## Code samples
+More plugin code snippets are available in [CustomAttestation.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md). The sample below is mainly for demonstrating the use of new APIs.
+
+The plugin-aware verifier application, which can either be the enclave or the host, can verify the evidence like this:
+
+`verifier.c`
+
+```C
+#include <verifier.h>
+
+/* Register verifiers upon initialization. */
+oe_verifier_t* verifiers;
+size_t verifiers_length;
+oe_get_verifier_plugins(&verifiers, &verifiers_length);
+for (size_t n = 0; n < verifiers_length; n++)
+    oe_register_verifier(&verifiers[n], NULL, 0);
+
+/* Build array of registered verifier format IDs. */
+_build_format_ids(verifiers, verifiers_length, &format_ids);
+
+/* Send the format IDs that can be verified to the attester.
+   Protocol is up to enclave and verifier. */
+send(VERIFIER_SOCKET_FD, format_ids, verifiers_length*sizeof(oe_uuid_t), 0);
+
+/* Receive a selected format ID. */
+recv(ENCLAVE_SOCKET_FD, &format_id, &format_id_size, 0);
+
+/* Get evidence parameters for the format ID. */
+oe_get_evidence_parameters(format_id, &params, &params_size);
+
+/* Send the evidence parameters to the attester */
+send(VERIFIER_SOCKET_FD, params, params_size, 0);
+
+/* Receive evidence and endorsement buffer from enclave. */
+recv(ENCLAVE_SOCKET_FD, evidence, evidence_size, 0);
+recv(ENCLAVE_SOCKET_FD, endorsements, endorsements_size, 0);
+
+/* Verify evidence. Can check the claims if desired. */
+oe_verify_evidence(
+    evidence,
+    evidence_size,
+    endorsements,
+    endorsements_size,
+    NULL,
+    0,
+    &claims,
+    &claims_size);
+
+/* Free data and unregister plugin. */
+oe_free_evidence_parameters(params);
+oe_free_format_ids(format_ids);
+oe_free_claims_list(claims, claims_size);
+for (size_t n = 0; n < verifiers_length; n++)
+    oe_unregister_verifier(&verifiers[n]);
+oe_free_verifier_plugins(verifiers);
+```
+
+The plugin-aware application enclave can generate the evidence using the plugin like this:
+
+`attester.c`
+
+```C
+#include <attester.h>
+
+/* Register attesters upon initialization. */
+oe_attester_t* attesters;
+size_t attesters_length;
+oe_get_attester_plugins(&attesters, &attesters_length);
+for (size_t n = 0; n < attesters_length; n++)
+    oe_register_attester(&attester[n], NULL, 0);
+
+/* Receive format IDs from verifier. Protocol is up to enclave and verifier. */
+recv(ENCLAVE_SOCKET_FD, &format_ids, &format_ids_size, 0);
+
+/* Selects a format ID from the received format IDs. */
+oe_select_attester_format(format_ids, format_ids_size/sizeof(oe_uuid_t), &selected_format);
+
+/* Send the selected format to attester */
```

Comment:
I think you mean "Send the selected format to the verifier"

Reply:
Fixed. Thanks.

---
In reply to: [396207707](https://github.com/openenclave/openenclave/pull/2621#discussion_r396207707) [](ancestors = 396207707)